CORE-8687 Track vassert messages for crash log

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This pull request adds the capability to the crash tracker to log out aborts caused by `vassert`.

* `vassert` messages are now captured in a thread local storage instance that will be accessed by
the `SIGABRT` signal handler.  This message will be written out to the generated crash report
* The `vassert` macro will now call `abort()` rather than `__builtin_trap()`.  This is because the implementation of `__builtin_trap()` is implementation dependent and may or may not raise a `SIGILL` signal.
* New debug mode only endpoints are added for testing purposes to set the value contained by the assert message storage class for testing

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
vassert: Created assert_log_holder

Created a structure that will be used to hold a callback function that
will be called when an assert is fired.  The desire will be for the
crash tracker module to register a callback that will trigger creation
of a crash tracker file containing the vassert message and stack trace.

A call back was used rather than directly calling into the crash tracker
from assert in order to not make crash tracker a global depenency, as
vassert is today.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
vassert: Use assert_log_holder in vassert

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
crash_tracker: Added handling for assertions

* Added assertions crash_type
* Registered callback with the assert message holder
* Callback will write out assert message to crash tracker file

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/json: Added routes for setting vassert message

New route in `/v1/debug/ctracker/va/{shard}` that permits setting an
assertion message in crash tracker (without calling vassert) from a
specified shard.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin: Implement /v1/debug/ctracker/va/{shard}

Implements handler that will trigger a crash tracker message creation
caused by vassert (without calling vassert).  Will verify that {shard}
is correct and take in a message of type {"message": <message>}.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin: Hook up access to ctracker/va for testing

This endpoint is only available in DEBUG builds of Redpanda.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt/admin: Added methods to /v1/debug/ctracker/va

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Added test for vassert messaging

Signed-off-by: Michael Boquard <michael@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61645-->
<!--dabb0cae-28d6-42c4-82cc-199b8772246e-->
<!--f8118bb7-21b2-4570-949a-49d0e3ac5fd1-->
<!--title-->
## Retry command for [Build#61645](https://buildkite.com/redpanda/redpanda/builds/61645)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/storage_failure_injection_test.py::StorageFailureInjectionTest.test_storage_failure_injection
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--f902fdfd-0809-4ceb-ae9d-b28cb5a49609-->
<!--TEST RESULTS-->
<!--601cb0b3-429b-4f96-8b19-f4615f78a581-->
<!--TEST RESULTS-->
<!--ec6d8acd-b579-4153-921a-82354a2b99db-->
<!--TEST RESULTS-->
<!--36141105-1e6f-486c-b344-3e6b599961fb-->
<!--TEST RESULTS-->
<!--f212652f-8846-449d-805c-350ad5564804-->
<!--TEST RESULTS-->
<!--e49b80d2-3a8d-4cf5-82e9-05eb6b86e9e9-->
<!--TEST RESULTS-->
<!--63863dbb-dd29-4aba-96db-cd49158adcdf-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61645">build#61645</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| assert_log_holder_test_rpunit.assert_log_holder_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61645#0194d8f6-2372-4a0e-ba8d-84c0f118a065 | FAIL | 0/2 |
| assert_log_holder_test_rpunit.assert_log_holder_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61645#0194d8f6-2373-411c-a317-0909013d80f2 | FAIL | 0/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61645#0194d954-7b74-4708-9428-2495e0f5876f | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61645#0194d954-7b74-4708-9428-2495e0f5876f | FLAKY | 1/2 |
| rptest.tests.storage_failure_injection_test.StorageFailureInjectionTest.test_storage_failure_injection | ducktape | https://buildkite.com/redpanda/redpanda/builds/61645#0194d938-77ec-4564-9024-c364ca23d33d | FAIL | 0/20 |
| rptest.tests.storage_failure_injection_test.StorageFailureInjectionTest.test_storage_failure_injection | ducktape | https://buildkite.com/redpanda/redpanda/builds/61645#0194d954-7b75-4de3-aa6f-724aba69073c | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61689">build#61689</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61689#0194dd28-3869-4b8b-984d-460f70d5cf39 | FLAKY | 1/3 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61689#0194dd2c-2abf-4d74-967d-9ca4d910b390 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61689#0194dd28-3869-483d-ab3e-70c2c8beb87d | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61689#0194dd2c-2abd-4b2c-91f4-f3b5f295937e | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/61689#0194dd2c-2abd-47b1-af84-366e29982458 | FLAKY | 1/4 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_cross_shard.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61689#0194dd2c-2abd-4b2c-91f4-f3b5f295937e | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61730">build#61730</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61730#0194e113-6369-4853-8da4-fbf4865b1ff5 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61730#0194e0f7-5783-4198-8ed7-90516a7ec14a | FLAKY | 1/2 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/61730#0194e113-6368-4a7b-8793-07cb1d10d25b | FLAKY | 1/5 |
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/61730#0194e113-6368-454a-a30e-7d261e447951 | FLAKY | 1/3 |
| rptest.tests.full_node_recovery_test.FullNodeRecoveryTest.test_node_recovery.recovery_type=full | ducktape | https://buildkite.com/redpanda/redpanda/builds/61730#0194e0f7-5781-4532-9b56-54f6ec3a1fb5 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62047">build#62047</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62047#01952179-6a2e-49b4-91f7-48c7b235566d | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62047#01952179-6a2e-452d-817c-a301be6fb2bf | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62047#01952176-37b9-4fac-bd43-57f313408e39 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62047#01952179-6a2e-49b4-91f7-48c7b235566d | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62047#01952179-6a2c-4d84-83f3-6c7f536ded3e | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62047#01952179-6a2c-4d84-83f3-6c7f536ded3e | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62064">build#62064</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62064#01952452-47ba-4727-b75e-05f4d45fa5ef | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62064#01952456-2988-4737-b7f9-d2197ba85f0a | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62064#01952456-2988-4737-b7f9-d2197ba85f0a | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62072">build#62072</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/62072#01952550-b74a-4157-a0bd-e920eb7e2fe2 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62072#01952550-b74b-4d46-9761-22b5b8232cd5 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/62072#01952565-8f41-4be8-8660-2849b8a6da60 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62072#01952565-8f41-4998-ad5a-af04c7e10628 | FLAKY | 1/2 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=False.server_cfg=300000.topic_cfg=None | ducktape | https://buildkite.com/redpanda/redpanda/builds/62072#01952565-8f41-4be8-8660-2849b8a6da60 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62104">build#62104</a></summary>


| test_id                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62104#019528d7-7bc4-4914-ba93-9ca1bcd7b45c | FLAKY         | 1/2      |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62104#019528db-5a03-405e-b307-9c7850d03e0b | FLAKY         | 1/2      |
</details>
<!--/body-->

pgellert commented on src/v/base/vassert.h:
```diff
@@ -99,6 +99,6 @@ static thread_local assert_log_holder g_assert_log_holder;
               #x,                                                              \
               ##args);                                                         \
-            __builtin_trap();                                                  \
+            abort();                                                           \
```
The reasoning here is not that SIGILL isn't catchable (in fact we handle it already) but that `__builtin_trap()` is compiler-implementation-dependent (ref: [gcc](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005ftrap), [clang](https://clang.llvm.org/docs/LanguageExtensions.html#builtin-trap)) and may or may not raise SIGILL. Instead, `abort()` reliably raises SIGABRT, which means that we know that all `vassert` crashes can be handled in the SIGABRT handler.

The only downside of this switch is that now we will end up printing a stacktrace twice on vassert crashes (once in the vassert macro and once in seastar's SIGABRT handler).

An alternative here would be to `raise(SIGILL)` instead, but `abort()` seems like a more standard way of terminating a process here (that's also how the seastar allocator terminates the process on OOM).

pgellert commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -678,6 +678,82 @@
             ]
+        },
+        {
+            "path": "/v1/debug/store_message/{shard}",
```
For Docs Team reviewers: These routes are only for testing and should not be documented.

pgellert commented on src/v/base/vassert.h:
```diff
@@ -17,11 +17,64 @@
+ * This class holds messages generated by asserts which will be fetched by the
+ * crash tracker when handling the signal generated by `vassert`.
+ */
+class assert_log_holder {
```
We should also try to cover recording the message in the `unreachable` macro in addition to `vassert`. It looks like we could just extract + reuse `assert_log_holder` in a follow-up PR, right?

michael-redpanda replied:
Oh actually that's a good point we can do that in this PR

pgellert commented on src/v/base/vassert.h:
```diff
@@ -17,11 +17,64 @@
+    // According to https://en.cppreference.com/w/c/program/signal, referring to
+    // objects with static or thread-local storage duration that is not
+    // lock-free atomic is UB.
+    static thread_local inline std::atomic<const char*> _abuffer{nullptr};
```
Could you please remind me again why this needs to be `static thread_local inline`? And it does look like this needs to be `static thread_local inline` because it doesn't work as a regular member variable based on my testing.

michael-redpanda replied:
Because we need it unique to each thread, otherwise if >1 thread vasserts at the same time (and it was declared say `static inline`) it would write to the same buffer

pgellert commented on src/v/crash_tracker/recorder.cc:
```diff
@@ -159,8 +160,14 @@ void recorder::record_crash_sighandler(recorded_signo signo) {
-        record_message(cd, "Aborting");
+        if (::detail::g_assert_log_holder.event_registered_signal_safe()) {
+            cd.type = crash_type::assertion;
+            record_message(
+              cd, ::detail::g_assert_log_holder.get_signal_safe_log());
```
Fyi, `record_message` here appends "on shard X" to the message. That ends up being useful, so I am not against it. Originally, I intended `record_message()` to be just a helper to format the signo-specific messages which are all like "_Something happened_ on shard X". This suggestion below is an alternative, but I am happy with the current output as well tbh:
```suggestion
            auto& format_buf = cd.crash_message;
            fmt::format_to_n(
              format_buf.begin(),
              format_buf.capacity(),
              "[shard {}] {}",
              ss::this_shard_id(),
              ::detail::g_assert_log_holder.get_signal_safe_log());
```

michael-redpanda replied:
I also like it informing which shard it occurred on.

pgellert reviewed (commented):
Looks great, mostly just questions from me. (+ fyi the new unit tests are failing)

michael-redpanda committed:
unreachable: Use g_assert_log_holder

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
> (+ fyi the new unit tests are failing)

Yeah the perils of having two build systems.  Forgot `GTEST` in `rp_unit`

michael-redpanda committed:
dt: Updated failure injection log allow list

Changes necessary due to changes in how vassert messaging is
constructed.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push [6bd5f0c](https://github.com/redpanda-data/redpanda/commit/6bd5f0cf881a1d9696efbaa8928dd2b121e77701):

* Added use of the assert storage class to `unreachable()`
* Fixed issue with tests CMake file
* Adjusted failure injection log allow list

dotnwat commented on src/v/base/vassert.h:
```diff
@@ -99,6 +99,6 @@ static thread_local assert_log_holder g_assert_log_holder;
               #x,                                                              \
               ##args);                                                         \
-            __builtin_trap();                                                  \
+            abort();                                                           \
```
* I might suggest `std::abort()` to avoid any potential for the wrong symbol being used here since this file is included pretty much everywhere in redpanda.

* are we at any risk for the same deadlock problems as glibc 2.26 (listed in the notes on the man page) related to the closing/flushing of IO streams, particularly since we are going to be doing I/O in response to sigabort?

* one of the original motivations behind builtin_trap was that it minimized the amount of code that was inlined into a vassert usage. presumably this isn't really an issue (at least in godbolt i see just a `call` instruction), but just noting that here for completeness.

> The only downside of this switch is that now we will end up printing a stacktrace twice on vassert crashes (once in the vassert macro and once in seastar's SIGABRT handler).

presumably this wouldn't be an issue if we hooked into Seastar's signal handling hooks? Then we could keep builtin_trap?

michael-redpanda replied:
> presumably this wouldn't be an issue if we hooked into Seastar's signal handling hooks? Then we could keep builtin_trap?

Our issue with builtin_trap is that what it does is implementation dependent.  In I think most cases it raises a `SIGILL` but not all.  We could change it to call `raise(SIGILL)` if we wish to maintain a similar behavior and get around any possible deadlocking issues caused by abort.

dotnwat replied:
let me rephrase. is there a benefit to

```
vassert():
logging
signal

signal_handler():
crash reporting
```

vs

```
vassert():
logging
crash reporting
builtin_trap
```

michael-redpanda replied:
talked with @pgellert about this and we believe the second approach is preferrable.  Less work in the handler is better, and avoids any annoying issues with memory of I/O getting messed up in the signal handler.

dotnwat commented on src/v/redpanda/admin/debug.cc:
```diff
@@ -501,6 +501,19 @@ void admin_server::register_debug_routes() {
           return get_node_uuid_handler();
       });
+
+    if constexpr (admin_server::is_store_message_enabled()) {
+        register_route<superuser>(
+          ss::httpd::debug_json::put_store_message,
+          [this](std::unique_ptr<ss::http::request> req) {
+              return put_store_message(std::move(req));
+          });
+        register_route<superuser>(
+          ss::httpd::debug_json::get_store_message,
+          [this](std::unique_ptr<ss::http::request> req) {
+              return get_store_message(std::move(req));
+          });
+    }
```
i think its worth debating if these are actually compiled out vs compiled in but disabled. for example compiling them in, and requiring some test-only runtime switch to make them active woudl mean we could test in both debug and release modes...

michael-redpanda replied:
I took inspiration from the failure injector stuff.  I agree that doing a runtime switch is preferable (we kind of do that with the `RP_FIXTURE_TEST` environment variable for the OpenSSL Context stuff in unit tests).  Happy to discuss a path forward on this and adapting it to other things like failure injector.

dotnwat replied:
it doesn't matter to me one way or the other

michael-redpanda replied:
https://redpandadata.atlassian.net/browse/CORE-9051

michael-redpanda replied:
Let's continue with this approach for now and get some eyes/consensus on 9051.  Once that's settled we'll come back and adjust.

michael-redpanda commented:
Force push [432d396](https://github.com/redpanda-data/redpanda/commit/432d396730b0212c2faf3e0b265d0f4096f02774):

* Rebase off dev to fix conflicts.  No changes

dotnwat commented on src/v/base/vassert.h:
```diff
@@ -17,11 +17,64 @@
+    // lock-free atomic is UB.
+    static thread_local inline std::atomic<const char*> _abuffer{nullptr};
+};
+static thread_local assert_log_holder g_assert_log_holder;
```
i'm concerned this isn't going to do what you want (or that i'm mistaken about what the exact goal is). here you're going to i think get a separate instance of `assert_log_holder` in each translation unit (`static foo` in a header--very rare what is intended), then those would all contain another thread_local `_abuffer` that points to a single instance (since its inline)? it's a bit of brain twister.

would this be simpler:

```
class assert_log_holder {
  std::atomic...;
};

inline thread_local assert_log_holder g_assert_log_holder
```

here you get a single global instance of `assert_log_holder` per thread.

michael-redpanda replied:
~~this is probably going to go away with the refactor to write out the file before raising the signal... this was only here so that the signal handle could access the vassert message~~

I re-read what you wrote and you're correct, the `static` declaration was wrong here.  The `thread_local` is also not going to be necessary moving forward with the new implementation.  Thanks for the call out.

michael-redpanda commented:
Note for reviewers: Need to reword some of the commits still, but wanted to get the changes up

Force push [a0b17db](https://github.com/redpanda-data/redpanda/commit/a0b17db5b4591fff6a6dacb7b3a906cec035ac7e):

* Fixed license in assert_log_holder_test.cc
* Major change: No longer deferring write to signal handler - writing to crash tracker file will now happen inline with vassert
  * Added a callback function to register with assert holder that will be used when message is formatted
  * Updated crash tracker to register a callback and write out message to crash tracker file
* Updated admin endpoints - no longer put/store message, now just a single put to trigger crash tracker file to write due to vassert (doesn't actually call vassert)
* No longer using `abort()` in unreachable or vassert - back to `__builtin_trap()`
* Updated tests

michael-redpanda commented:
Force push [a0b17db](https://github.com/redpanda-data/redpanda/commit/a0b17db5b4591fff6a6dacb7b3a906cec035ac7e):

* Fixed lint error in python

pgellert commented on src/v/base/vassert.h:
```diff
@@ -16,12 +16,61 @@
 #include <seastar/util/backtrace.hh>
 #include <seastar/util/log.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+#include <atomic>
```
nit: no atomics are used at the moment
```suggestion
```

michael-redpanda replied:
Ha ha, Gellert of the past.  We are using atomics again!

pgellert commented on tests/rptest/tests/crash_loop_checks_test.py:
```diff
@@ -286,3 +292,33 @@ def signo_prefix():
+                                      node=self.broker)
+
+        self.redpanda.signal_redpanda(self.broker,
+                                      signal=signal.SIGABRT,
```
nit: Technically, if we want to mirror the `vassert` implementation, this will be a SIGILL because that's what `__builtin_trap()` leads to. But that's a minor detail. In fact, for this test, we wouldn't even need to send a signal to the process, and the crash report would still have been generated, right?
```suggestion
                                      signal=signal.SIGILL,
```

michael-redpanda replied:
> nit: Technically, if we want to mirror the vassert implementation, this will be a SIGILL because that's what __builtin_trap() leads to

Yeah good call I'll change that

> In fact, for this test, we wouldn't even need to send a signal to the process, and the crash report would still have been generated, right?

Part of what's being tested here is ensuring that it's the assert message being written to the crash report and not the SIGILL/SIGABRT message being written by the signal handler.

michael-redpanda commented:
Force push [764a2f0](https://github.com/redpanda-data/redpanda/commit/764a2f06a03b0e4d271d62e771a62803cd2c1f68):

* No code changes - just commit rewording

pgellert commented on src/v/base/vassert.h:
```diff
@@ -16,12 +16,61 @@
+private:
+    std::optional<assert_cb_func> _cb_func;
+};
+inline assert_log_holder g_assert_log_holder;
```
I think we need to make `g_assert_log_holder` or `_cb_func` either `thread_local` or use some synchronization around it because at the moment there is a data race, for example, if `register_cb()` is called on T0 and `register_event()` on T1. (It's an unlikely race, but still.)

pgellert commented on src/v/base/tests/BUILD:
```diff
@@ -1,4 +1,4 @@
-load("//bazel:test.bzl", "redpanda_cc_btest", "redpanda_cc_btest_no_seastar")
+load("//bazel:test.bzl", "redpanda_cc_btest", "redpanda_cc_btest_no_seastar", "redpanda_cc_gtest")
```
nit: This and some other commit messages still talk about SIGABRT, which is no longer relevant

pgellert commented on src/v/base/unreachable.h:
```diff
@@ -19,9 +19,10 @@
+          "This code should not be reached ({}:{})",                           \
+          __FILE__,                                                            \
+          __LINE__);                                                           \
         __builtin_trap();                                                      \
```
nit: remove ` and abort` from this commit message

pgellert commented on tests/rptest/services/redpanda.py:
```diff
@@ -213,7 +213,9 @@ class CloudStorageCleanupStrategy(enum.Enum):
     re.compile("assert - Backtrace below:"),
-    re.compile("finject - .* flush called concurrently with other operations")
+    re.compile("finject - .* flush called concurrently with other operations"),
+    re.compile("assert - Backtrace:"),
+    re.compile("Aborting on shard")
```
* I think line 217 replaces line 215
* I think 218 is not necessary now that we're using `__builtin_trap()` again. Am I missing something?

pgellert reviewed (commented):
Looks great

michael-redpanda commented:
Force push [783a803](https://github.com/redpanda-data/redpanda/commit/783a803cc58decf49f0e250c961109a2aa38c414):

* Fixed up commit messages to remove references to SIGABRT (no longer using abort() in vassert/unreachable)
* Switched DT test to use SIGILL rather than SIGABRT
* Removed unecessary failure log messages in redpanda.py
* Hardened callback mechanism to protect against a race condition that may occur if one thread is registering a callback while another is attempting to register an event

dotnwat commented on src/v/base/vassert.h:
```diff
@@ -16,12 +16,63 @@
+    }
+
+    void register_cb(assert_cb_func cb) {
+        assert_cb_func before = nullptr;
+        _cb_func.compare_exchange_strong(before, cb);
```
do you want first writer wins, or last writer wins?

michael-redpanda replied:
If by "writer" you mean whoever calls `register_cb`, I'd say probably first... but in practice there's only one spot where this is going to be called (from recorder.cc)

dotnwat replied:
in that case you could make a stronger guarantee and assert the pointer is null when its set.

dotnwat commented on src/v/base/vassert.h:
```diff
@@ -16,12 +16,63 @@
+            cb_func(buffer);
+        }
+
+        g_assert_log.l.error("{}", buffer);
+        g_assert_log.l.error("Backtrace:\n{}", bt);
```
should the log lines take precedence over the callback for example in the case the callback has issues given the state of the system?

michael-redpanda replied:
Yeah that's a very good point, thanks.

dotnwat commented on src/v/redpanda/admin/debug.cc:
```diff
@@ -1029,3 +1032,43 @@ ss::future<ss::json::json_return_type> admin_server::override_node_uuid_handler(
     co_return ss::json::json_return_type(ss::json::json_void());
 }
+
+ss::future<std::unique_ptr<ss::http::reply>> admin_server::put_ctracker_va(
```
nit: should it go into the finjector url namespace?

michael-redpanda replied:
Unless i'm completely blind I don't see a `finjector` namespace.  There is `failure-probes` but that's all honey badger related, including the root `/v1/failure-probes` endpoint.

dotnwat replied:
isn't this a failure probe?

michael-redpanda replied:
It's not associated with honey-badger.  Everything in `/v1/failure-probes` connects with honey badger

dotnwat replied:
it can be both only containing honey badger bits, and also a category of end points.

michael-redpanda commented:
Force push [3b13263](https://github.com/redpanda-data/redpanda/commit/3b13263f2bdc3ddf61432c3816c9c99b2d0a6108):

* Moved logging of assert message to before callback

pgellert commented on src/v/base/vassert.h:
```diff
@@ -16,12 +16,63 @@
 
 #include <seastar/util/backtrace.hh>
 #include <seastar/util/log.hh>
+#include <seastar/util/noncopyable_function.hh>
```
nit:
```suggestion
```

(I can carry this over to a follow up PR.)

pgellert reviewed (approved):
Looks good. (There's just one remaining unused header in vassert.h. I am happy to clean that up in one of my follow up PRs.)
